### PR TITLE
Make tuple types more round-trippable.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Serialize the information so we can call the tuple-factory-method that doesn't need an underlying type.  We need this when the underlying type is an error type (for example, when there is no reference to System.ValueTuple).  The factory-method that takes an underlying type throws if the underlying type is an error type.

This is necessary to support hte invariant that types can be round-tripped through SymbolKey, and this was found while testing FAR in OOP.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide @jcouv 

@jcouv This 'fixes' the issue... however, i'm concerned the issue exists in the first place.  Here's the problem i have with the existing API: If you have a tuple in code, but no reference to System.ValueTuple, then you'll get a TupleTypeSymbol with an ErrorType for it's UnderlyingTupleType.  This makes sense, but causes a problem later on.  Specifically, if you then try to create a TupleTypeSymbol using that underlying-tuple-type, the compiler will throw.

I feel like the compiler should be relaxed here.  If i have a TupleTypeSymbol, then i should be able to recreate it using its UnderlyingTupleType.  What do you think?

**Comment by jcouv** [comment]
@CyrusNajmabadi Just to confirm my understanding: you're suggesting that CreateTupleTypeSymbol (the one with an underlying type) should accept an error type as the underlying type?
I wonder if we should keep it tighter than that: maybe only if the error type has the proper name. What do you think?

File: src/Workspaces/Core/Portable/SymbolKey/SymbolKey.TupleTypeSymbolKey.cs [hunk 1]
```diff
@@ -13,37 +13,43 @@ private static class TupleTypeSymbolKey
             public static void Create(INamedTypeSymbol symbol, SymbolKeyWriter visitor)
             {
                 Debug.Assert(symbol.IsTupleType);
-                visitor.WriteSymbolKey(symbol.TupleUnderlyingType);
 
-                var friendlyNames = ArrayBuilder<String>.GetInstance();
+                var elementTypes = ArrayBuilder<ISymbol>.GetInstance();
+                var friendlyNames = ArrayBuilder<string>.GetInstance();
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/SymbolKey/SymbolKey.TupleTypeSymbolKey.cs]: Should pass counts to `GetInstance` for efficiency.

File: src/Workspaces/Core/Portable/SymbolKey/SymbolKey.TupleTypeSymbolKey.cs [hunk 2]
```diff
@@ -13,37 +13,43 @@ private static class TupleTypeSymbolKey
...
-                {
-                    return new SymbolKeyResolution(reader.Compilation.ObjectType);
+                    try
+                    {
+                        var result = reader.Compilation.CreateTupleTypeSymbol(
+                            elementTypes, elementNames, elementLocations);
+                        return new SymbolKeyResolution(result);
+                    }
+                    catch (ArgumentException)
+                    {
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/SymbolKey/SymbolKey.TupleTypeSymbolKey.cs]: Can we add a comment as to why this exception is okay?  Not obvious at a glance.

**Comment by CyrusNajmabadi** [comment]
@jcouv I'm not sure i have an exact suggestion yet.  However, it seems very strange to me that if i get a TupleType back from teh compiler, and that TupleType has an underlying type, that the compiler then throws if i try to make that TupleType again with that same underlying type. :)

I think your suggestion seems reasonable.  An error type called "System.ValueTuple" should be acceptable since it's really just a NamedType, albeit one that we couldn't resolve from source or metadata.

File: src/Workspaces/Core/Portable/SymbolKey/SymbolKey.TupleTypeSymbolKey.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/SymbolKey/SymbolKey.TupleTypeSymbolKey.cs]: It's part of the contract of CreateTupleType.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/SymbolKey/SymbolKey.TupleTypeSymbolKey.cs]: but i will comment.

**Comment by jcouv** [comment]
@CyrusNajmabadi Agreed. Feel free to file an issue.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Need reviews plz

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Need reviews plz  This is blocking enabling FAR and navigateTo OOP.

**Comment by Pilchie** [comment]
@CyrusNajmabadi We're in escrow for master and rc-3 now.  Can you fill out the ask mode info for this?

**Comment by Pilchie** [comment]
(Also, we should probably apply this change to `dev15-rc3` as well.

**Comment by CyrusNajmabadi** [comment]
what is the difference between dev15-rc3 and master?

**Comment by CyrusNajmabadi** [comment]
**Customer scenario**

Customer performs a FAR operation on a tuple-element, without a reference to System.ValueTuple.dll.  This will crash if OOP is turned on and FAR is running in OOP.  As such, without this fix, this blocks us from turning on OOP for FAR.

**Bugs this fixes:** 

None.

**Workarounds, if any**

We do not turn on OOP for FAR.

**Risk**

Very low risk.

**Performance impact**

Very low impact.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

We did not miss this.  our tests caught the issue when we wanted to turn on OOP for FAR. 

**How was the bug found?**

Developer work.

--

Tagging @MattGertz

**Comment by CyrusNajmabadi** [comment]
@MattGertz I jumped the gun on merging.  Sorry about that.

**Comment by MattGertz** [comment]
@CyrusNajmabadi @Pilchie I've set this for RTM, since I'm guessing that we're not turning on FAR & OOP at the last second for RC3, but let's discuss if you disagree.

**Pull request merged**